solver/src/settlement_submission/retry.rs

```diff
@@ -0,0 +1,136 @@
+use super::{EncodedSettlement, GAS_PRICE_REFRESH_INTERVAL, MAX_GAS};
+use anyhow::Result;
+use contracts::GPv2Settlement;
+use ethcontract::{
+    errors::{ExecutionError, MethodError},
+    jsonrpc::types::Error as RpcError,
+    transaction::{confirm::ConfirmParams, ResolveCondition},
+    web3::error::Error as Web3Error,
+    GasPrice,
+};
+use futures::{stream, Stream, StreamExt};
+use gas_estimation::GasPriceEstimating;
+use primitive_types::U256;
+use std::time::Duration;
+use transaction_retry::{TransactionResult, TransactionSending};
+
+// Create a never ending stream of gas prices based on checking the estimator in fixed intervals
+// and enforcing the minimum increase. Errors are ignored.
+pub fn gas_price_stream(
+    target_confirm_time: Duration,
+    gas_price_cap: f64,
+    estimator: &dyn GasPriceEstimating,
+) -> impl Stream<Item = f64> + '_ {
+    let stream = stream::unfold(true, move |first_call| async move {
+        if !first_call {
+            tokio::time::delay_for(GAS_PRICE_REFRESH_INTERVAL).await;
+        }
+        let estimate = estimator
+            .estimate_with_limits(MAX_GAS as f64, target_confirm_time)
+            .await;
+        Some((estimate, false))
+    })
+    .filter_map(|gas_price_result| async move {
+        match gas_price_result {
+            Ok(gas_price) => {
+                tracing::debug!("estimated gas price {}", gas_price);
+                Some(gas_price)
+            }
+            Err(err) => {
+                tracing::error!("gas price estimation failed: {:?}", err);
+                None
+            }
+        }
+    });
+    transaction_retry::gas_price_increase::enforce_minimum_increase_and_cap(gas_price_cap, stream)
```

Comment:
I personally think it would be nice to expose a gas price stream factory that takes a target time, amount, price_cap, estimator and a "sleep" function from the retry or price estimating crate. It seems generic enough to me and the logic with the unfolded stream is a bit nifty (so nice not to maintain it twice).

Reply:
The problem is that there are too many implementation details like sleeping, logging and how to call the estimator and there is no good generic way to take them. To me duplicating makes more sense.

Reply:
This discussion goes outside the scope of this PR, but what's the concern of making the retry library opinionated on e.g. logging/sleeping (not that we could pass it as abstract traits from above) similar to how other libraries (e.g. web3 are)? 

It find there is still quite a lot of manual code involved to integrate transaction retries and it feels to me that the use case of wanting to submit some MethodBuilder with a specific target time and a gas price estimator is quite common.

Reply:
I suspect that it wouldn't be useful because people want to customize too many parts of it. Like what if you want to do some extra work every time a resubmission happens other than what it provides by default. I did make an issue now on the retry repo to try implementing this with an optional `ethcontract` feature so I am going to try it out.
I have another idea for the gas price stream too.